function-web/src/main/java/com/future/function/web/controller/communication/chatroom/ChatroomController.java

```diff
@@ -99,14 +99,14 @@ public ChatroomController(
       return ChatroomResponseMapper.toPagingChatroomResponse(
         chatroomService.getChatroomsWithKeyword(
           search, session.getUserId(), PageHelper.toPageable(page, size)),
-        messageService, messageStatusService, fileProperties.getUrlPrefix(),
+        messageService, messageStatusService, userService, resourceService, fileProperties.getUrlPrefix(),
```

Comment:
I do not prefer passing service to response mapper, especially when in response mapper, the mapper calls the service.

I think it's best to use `userService` and `resourceService` in `chatroomService.getChatroomsWithKeyboard` method implementation, and return a tuple of the group picture object and the user object itself with the chatroom object.

Reply:
i still don't get the reason why it is better to not passing service to response mapper, but if i do and then moved userService and resourceService, in the end i still can't move mesageService and messageStatusService though because i need the data from both services in the response. And i don't make any change to both services in this PR so it is already merged to master long time ago.

Reply:
I think it's about the task of each layer, and in this case, I believe response mapper layer should only map database objects or other objects to web response objects. For such reason I prefer to not pass service objects to response mapper as doing so will cause response mapper to do something out of its responsibility, such as database object retrieval from repository through service.

Maybe the object retrieval (say in this case the File object retrieval from database) could be done in ChatroomServiceImpl instead, and is returned to controller with using Pair (from Spring Framework) or Tuple (currently not available in project) of both Chatroom and File object in one Pair/Tuple object, then pass the retrieved object from chatroomService to the response mapper.

If such modification is too large, at least (though I don't quite prefer this) maybe we can call other services from the controller instead of from the response mapper.

Or if any other ideas are available, let's discuss them here. Any ideas as well @oliverrsebastian @RickyKennedy25 ?